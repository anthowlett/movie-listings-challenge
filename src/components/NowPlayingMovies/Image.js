import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

// Styles
const ImageHolder = styled.div`
  background: #000 none no-repeat top center;
  background-image: url('imgs/cover-placeholder.gif');
  background-size: cover;
  border: 1px solid #222;
`;

const Cover = styled.div`
  padding-bottom: 150%;
  ${props => props.bg && css`
    background: transparent none no-repeat top center;
    background-image: url(${props.bg});
  `};
  background-size: cover;
`;

export class Image extends Component {
    getImageSrc(name, base, size) {
      return base + size + name;
    }

    render() {
      const { config } = this.props;
      const size = config.poster_sizes[4];
      const src = this.getImageSrc(this.props.name, config.secure_base_url, size);
      
      return (
        <ImageHolder>
          <Cover bg={src} />
        </ImageHolder>
      );
    }
}

const mapStateToProps = (state) => ({
	config: state.config.images
});

Image.propTypes = {
  config: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  alt: PropTypes.string
};

export default connect(mapStateToProps)(Image);