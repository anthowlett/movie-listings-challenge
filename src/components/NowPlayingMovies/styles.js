import styled from 'styled-components';
import { breakpoints, colors } from '../../utils';

export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const MovieWrapper = styled.div`
  flex-basis: 100%;
  
  ${breakpoints.mini(`
      flex-basis: 50%;
  `)};
  ${breakpoints.small(`
      flex-basis: 33.333%;
  `)};
  ${breakpoints.medium(`
      flex-basis: 25%;
  `)};
`;

export const MovieDetails = styled.div`
  padding: 5px 15px;
`;

export const Message = styled.p`
  width: 100%;
  padding: 20px;
  text-align: center;
  font-size: 20px;
  color: red;
  background-color: ${colors.black_5};
`;