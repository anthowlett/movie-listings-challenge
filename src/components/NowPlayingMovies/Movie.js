import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { breakpoints } from '../../utils';
import Image from './Image';

// Styles
const Wrapper = styled.div`
  flex-basis: 100%;
  
  ${breakpoints.mini(`
      flex-basis: 50%;
  `)};
  ${breakpoints.small(`
      flex-basis: 33.333%;
  `)};
  ${breakpoints.medium(`
      flex-basis: 25%;
  `)};
`;

const Details = styled.div`
  padding: 5px 15px;
`;

const Movie = ({
  data,
  genres
}) => (
	<Wrapper>
		<Image name={data.poster_path} alt={data.title} />
		<Details>
			<h2>{data.title}</h2>
			<div>
				<p>Rated <strong>{data.vote_average}</strong> / 10</p>
				<p>Popularity <strong>{Math.round(data.popularity)}</strong></p>
        <ul>
          {
            genres
              .filter(genre => data.genre_ids.includes(genre.id))
              .map((genre) => (
                <li key={genre.id}>{genre.name}</li>
              ))
          }
        </ul>
			</div>
		</Details>
	</Wrapper>
);

Movie.propTypes = {
  genres: PropTypes.array.isRequired,
  data: PropTypes.object.isRequired
};

export default Movie;