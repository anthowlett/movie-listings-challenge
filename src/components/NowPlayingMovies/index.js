import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { filters } from '../../utils';
import Movie from './Movie';
import { colors } from '../../utils';

// Styles
export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Message = styled.p`
  width: 100%;
  padding: 20px;
  text-align: center;
  font-size: 20px;
  color: red;
  background-color: ${colors.black_5};
`;

class NowPlayingMovies extends Component {
  render() {
    const { movies, genres } = this.props;
    
    return (
      <Container>
        {
          !movies.length ?
            <Message>Sorry friend, we've got nothing :(</Message> : 
            movies.map((item) => (
              <Movie key={ item.id } data={ item } genres={ genres } />
            ))
        }
      </Container>
    );
  }
}

const mapStateToProps = (state) => ({
	genres: state.genres.list,
	movies: filters.filterMovies(state)
});

NowPlayingMovies.propTypes = {
  genres: PropTypes.array.isRequired,
  movies: PropTypes.array.isRequired
};

export default connect(mapStateToProps)(NowPlayingMovies);