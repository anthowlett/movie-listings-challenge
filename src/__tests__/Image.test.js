import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Image } from '../components/NowPlayingMovies/Image';
import '../setupTests';

describe('NowPlayingMovies', () => {
  
  describe('<Image />', () => {
    const data = {
      poster_path: '/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg',
      title: 'Aquaman'
    };
    const config = {
      poster_sizes: ['w92', 'w154', 'w185', 'w342', 'w500', 'w780', 'original'],
      secure_base_url: 'https://image.tmdb.org/t/p/'
    }
    const tree = renderer.create(
      <Image name={data.poster_path} alt={data.title} config={config} />
    );

    it('should render', () => {
      expect(tree.toJSON())
        .toMatchSnapshot();
    });

    it('should contain a placeholder image', () => {
      expect(tree.toJSON())
        .toHaveStyleRule('background-image', 'url(\'imgs/cover-placeholder.gif\')');
    });

    it('should contain a cover image', () => {
      expect(tree.toJSON().children[0])
        .toHaveStyleRule('background-image', 'url(https://image.tmdb.org/t/p/w500/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg)');
    });

    it('should only have one child', () => {
      expect(tree.toJSON().children.length)
        .toEqual(1);
    });
  });

});