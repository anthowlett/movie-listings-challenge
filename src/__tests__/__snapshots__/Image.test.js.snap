// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NowPlayingMovies <Image /> should render 1`] = `
.c0 {
  background: #000 none no-repeat top center;
  background-image: url('imgs/cover-placeholder.gif');
  background-size: cover;
  border: 1px solid #222;
}

.c1 {
  padding-bottom: 150%;
  background: transparent none no-repeat top center;
  background-image: url(https://image.tmdb.org/t/p/w500/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg);
  background-size: cover;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
</div>
`;
